File: csharp/Patterns/Adapter.cs
using System;

namespace DesignPatterns.Patterns;

public static class Adapter
{
    private interface IRoundPeg
    {
        double Radius { get; }
    }

    private class RoundPeg : IRoundPeg
    {
        public RoundPeg(double radius)
        {
            Radius = radius;
        }

        public double Radius { get; }
    }

    private class RoundHole
    {
        public RoundHole(double radius)
        {
            Radius = radius;
        }

        public double Radius { get; }

        public bool Fits(IRoundPeg peg) => peg.Radius <= Radius;
    }

    private class SquarePeg
    {
        public SquarePeg(double width)
        {
            Width = width;
        }

        public double Width { get; }
    }

    private class SquarePegAdapter : IRoundPeg
    {
        private readonly SquarePeg _peg;

        public SquarePegAdapter(SquarePeg peg)
        {
            _peg = peg;
        }

        public double Radius => _peg.Width * Math.Sqrt(2) / 2;
    }

    public static void Run()
    {
        Console.WriteLine("\nAdapter");
        var hole = new RoundHole(5);
        var smallSquare = new SquarePeg(5);
        var bigSquare = new SquarePeg(10);

        Console.WriteLine($"Small square fits: {hole.Fits(new SquarePegAdapter(smallSquare))}");
        Console.WriteLine($"Big square fits: {hole.Fits(new SquarePegAdapter(bigSquare))}");
    }
}

File: python/patterns/adapter.py
import math


class RoundPeg:
    def __init__(self, radius: float) -> None:
        self.radius = radius


class RoundHole:
    def __init__(self, radius: float) -> None:
        self.radius = radius

    def fits(self, peg: RoundPeg) -> bool:
        return peg.radius <= self.radius


class SquarePeg:
    def __init__(self, width: float) -> None:
        self.width = width


class SquarePegAdapter(RoundPeg):
    def __init__(self, peg: SquarePeg) -> None:
        self.peg = peg
        super().__init__(self._radius())

    def _radius(self) -> float:
        return self.peg.width * math.sqrt(2) / 2


def run() -> None:
    print("\nAdapter")
    hole = RoundHole(5)
    small_square = SquarePeg(5)
    big_square = SquarePeg(10)

    print("Small square fits:", hole.fits(SquarePegAdapter(small_square)))
    print("Big square fits:", hole.fits(SquarePegAdapter(big_square)))
